src/groupeddataframe/groupeddataframe.jl

```diff
@@ -410,3 +399,95 @@ function Base.get(gd::GroupedDataFrame, key::Union{Tuple, NamedTuple}, default)
         return default
     end
 end
+
+
+#
+# Indexing with arrays
+#
+
+# Array of integers OR bools (basis for other methods taking an array)
+function Base.getindex(gd::GroupedDataFrame, idxs::AbstractVector{<:Integer})
+    new_starts = gd.starts[idxs]
+    new_ends = gd.ends[idxs]
+    if !allunique(new_starts)
+        throw(ArgumentError("duplicates in idxs argument are not allowed"))
+    end
+    new_groups = zeros(Int, length(gd.groups))
+    idx = gd.idx
+    for i in eachindex(new_starts)
+        @inbounds for j in new_starts[i]:new_ends[i]
+            new_groups[idx[j]] = i
+        end
+    end
+    GroupedDataFrame(gd.parent, gd.cols, new_groups, gd.idx,
+                     new_starts, new_ends, length(new_starts))
+end
+
+# Colon (creates copy)
+Base.getindex(gd::GroupedDataFrame, idxs::Colon) =
+    GroupedDataFrame(gd.parent, gd.cols, gd.groups, gd.idx,
+                     gd.starts, gd.ends, gd.ngroups)
+
+# Convert an array of eltype other than <: Integer to an array of integer indices
+_to_index_array(gd::GroupedDataFrame, idxs::AbstractVector{<:GroupKey}) = [Base.to_index(gd, k) for k in idxs]
+_to_index_array(gd::GroupedDataFrame, idxs::AbstractVector{<:Tuple}) = [Base.to_index(gd, k) for k in idxs]
+_to_index_array(gd::GroupedDataFrame, idxs::AbstractVector{<:NamedTuple}) = [Base.to_index(gd, k) for k in idxs]
+
+# Non-definite eltype, infer and enforce homogeneity
+function _to_index_array(gd::GroupedDataFrame, idxs::AbstractVector)
+    # Edge case - array is empty
+    isempty(idxs) && return Int[]
+
+    # Infer eltype based on type of first index
+    idx1 = idxs[1]
+    T1 = typeof(idx1)
+
+    T = if T1 <: Integer && T1 !== Bool
+        Integer
+    elseif T1 <: GroupKey
+        GroupKey
+    elseif T1 <: Tuple
+        Tuple
+    elseif T1 <: NamedTuple
+        NamedTuple
+    else
+        throw(ArgumentError("Invalid index: $idx1 of type $T1"))
+    end
+
+    ints = zeros(Int, length(idxs))
+    for (i, idx) in enumerate(idxs)
+        idx isa GroupIndexTypes || throw(ArgumentError("Invalid index: $idx1 of type $T1"))
+        idx isa T || throw(ArgumentError("Mixed index types in array not allowed"))
```

Comment:
mixing integer types except `Bool` is allowed in Base, so we should also allow it here:
```
julia> x = [1,2,3]
3-element Array{Int64,1}:
 1
 2
 3

julia> x[Any[1,0x2]]
2-element Array{Int64,1}:
 1
 2
```